lightning/src/chain/chainmonitor.rs

```diff
@@ -82,18 +82,40 @@ where C::Target: chain::Filter,
 	/// descendants of such transactions. It is not necessary to re-fetch the block to obtain
 	/// updated `txdata`.
 	pub fn block_connected(&self, header: &BlockHeader, txdata: &TransactionData, height: u32) {
+		let mut dependent_txdata = Vec::new();
 		let monitors = self.monitors.read().unwrap();
 		for monitor in monitors.values() {
 			let mut txn_outputs = monitor.block_connected(header, txdata, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
 
+			// Register any new outputs with the chain source for filtering, storing any dependent
+			// transactions from within the block that previously had not been included in txdata.
 			if let Some(ref chain_source) = self.chain_source {
+				let block_hash = header.block_hash();
 				for (txid, outputs) in txn_outputs.drain(..) {
 					for (idx, output) in outputs.iter() {
-						chain_source.register_output(&OutPoint { txid, index: *idx as u16 }, &output.script_pubkey);
+						// Register any new outputs with the chain source for filtering and recurse
+						// if it indicates that there are dependent transactions within the block
+						// that had not been previously included in txdata.
+						let output = WatchedOutput {
+							block_hash: Some(block_hash),
+							outpoint: OutPoint { txid, index: *idx as u16 },
+							script_pubkey: output.script_pubkey.clone(),
+						};
+						if let Some(tx) = chain_source.register_output(output) {
+							dependent_txdata.push(tx);
+						}
 					}
 				}
 			}
 		}
+
+		// Recursively call for any dependent transactions that were identified by the chain source.
+		if !dependent_txdata.is_empty() {
+			dependent_txdata.sort_unstable_by_key(|(index, _tx)| *index);
+			dependent_txdata.dedup_by_key(|(index, _tx)| *index);
```

Comment:
> IIRC, `ChannelMonitor::block_connected` will never return a registration request for outputs spent by transactions it has just been called to parse ?
> 

I'm not sure I follow. Could you explain why it will not do so? If fact, doesn't my test demonstrate that it does? (i.e., a commitment transaction is processed and its outputs are registered)

> From where duplication could have been introduced ? Though I agree to keep it as belt-and-suspender.

The duplication may occur across channel monitors. For instance, two channel monitors register two different outputs which happen to be spent by the same in-block descendent transaction. So each call to `register_output` should return this transaction, which would need to be de-duped.

Would this ever happen in practice? Maybe not? But it is not an impossible scenario, right?

Reply:
> If fact, doesn't my test demonstrate that it does? (i.e., a commitment transaction is processed and its outputs are registered)

I meant a channel monitor won't ask to register outputs _spent_ by the processed commitment transactions, only outputs for the next stage of spending. Those outputs _spent_ are already registered, otherwise we won't have received a transaction spending them.

> Would this ever happen in practice? Maybe not? But it is not an impossible scenario, right?

I don't believe this can happen right now because we don't coalesce justice transactions across channels. Although  thought about doing it for fee optimizations a while back but not worthy the complexity.

Beyond, I'm also pretty sure that dual-funding/splicing will enable that kind of scenario where channel monitors register different outputs which happen to be spent by the same in-block descendent transaction. A splicing transaction could lawfully spend two funding outpoints.

AFAICT, I think those scenarios aren't possible for now, but really likely in the future so better to keep the dedup ?

Reply:
> I meant a channel monitor won't ask to register outputs _spent_ by the processed commitment transactions, only outputs for the next stage of spending ? Those outputs _spent_ are already registered, otherwise we won't have received a transaction spending them.

Ah, right. Sorry, I misread your message.
 
> I don't believe this can happen right now because we don't coalesce justice transactions across channels. Although thought about doing it for fee optimizations a while back but not worthy the complexity.
> 
> Beyond, I'm also pretty sure that dual-funding/splicing will enable that kind of scenario where channel monitors register different outputs which happen to be spent by the same in-block descendent transaction. A splicing transaction could lawfully spend two funding outpoints.
> 
> AFAICT, I think those scenarios aren't possible for now, but really likely in the future so better to keep the dedup ?

Gotcha, will keep as is. Though note that these transactions are published by the counterparty, which may be another implementation.

For a de-dup test, I suppose if I set up two channels then I may be able to form a transaction that spends from both commitment transaction's outputs.

Reply:
> For a de-dup test, I suppose if I set up two channels then I may be able to form a transaction that spends from both commitment transaction's outputs.

Actually, I think one channel would be sufficient here. A transaction spending the commitment transaction's `to_local` output and the HTLC transaction's output should suffice. The Electrum client would return this twice (once for each call to `register_output`), which would trigger the de-dup logic.

Is there an easy way to form such a transaction using our API?